apps/credence/credence.1
.Dd April 16, 2022
.Dt credence 1
.Os mcplex.net
.Sh NAME
.Nm credence
.Nd Generate or check key pairs for services using libDwmCredence
.Sh SYNOPSIS
.Nm
.Cm keygen
.Op Fl i Ar identity
.Op Fl d Ar directory
.Nm
.Cm keycheck
.Op Fl d Ar directory
.Nm
.Cm -v
.Sh DESCRIPTION
.Nm
is a simple command-line utility for creating and checking
public/private keypairs (Ed25519) to be used by services that
utilize libDwmCredence for authentication.
It operates in a somewhat similar manner to
.Xr ssh-keygen 1 but does not use passphrases and only uses Ed25519 keys
(other key types are not supported).
.Pp
.Xr credence 1 operates in two possible modes: key generation and key checking.
.Ss Key generation
.Nm
.Cm keygen
.Op Fl i Ar identity
.Op Fl d Ar directory
.Pp
Generates public and private key files.
The following command line options are available:
.Bl -tag -width indent
.It Fl i Ar identity
Specify the identity.
If this option is not used, a default of username@hostname will be used,
where \fIusername\fR will be determined via
.Xr getuid 2 and
.Xr getpwuid 2 and \fIhostname\fR will be  determined via
.Xr gethostname 3 .
.It Fl d Ar directory
Specify the directory in which the keys should be stored.
The public key will be stored in 'id_ed25519.pub' and the private key will be
stored in 'id_ed25519' in the given directory.
If this option is not used, ~/.credence will be used as the default directory.
.El
.Pp
A user will normally use their email address as the \fIidentity\fR, and not
use the \fI-d directory\fR argument.
They may then copy the contents of their \fI~/.credence\fR directory onto any
host from which they need access to services using libDwmCredence, taking care
to maintain 0600 permissions on their key files and 0700 permissions on their
\fI~/.credence\fR directory.
For example:
.Bd -literal
% credence keygen -i dwm@mcplex.net
.Ed
.Pp
A service will normally use a service name and hostname as the \fIidentity\fR,
and an appropriate location for the key files.
For example:
.Bd -literal
# credence keygen -i mcblockd@gw.mcplex.net -d /usr/local/etc/mcblockd
.Ed
.Ss Key checking
.Nm
.Cm keycheck
.Op Fl d Ar directory
.Pp
Checks the validity of a public/private key pair.
The following command line options are available:
.Bl -tag -width indent
.It Fl d Ar directory
Specify the directory in which the keys are stored.
If this option is not used, the default ~/.credence directory is used.
.El
.Pp
If the key pair is valid, "Valid key stash '<directory>'" will be printed
on stdout and the
.Xr credence 1
process will exit with status 0.
If any error occurs (invalid keypair, missing file(s), etc.),
.Xr credence 1
will print an error on stderr and exit with status 1.
.Sh FILES
.Bl -tag -width indent
.It Pa ${HOME}/.credence/id_ed25519
The user's credence private key file, created with
.Xr credence 1 .
This file should be owned by the user and have permissions 0600.
It must contain the private part of an Ed25519 key pair.
.It Pa ${HOME}/.credence/id_ed25519.pub
The user's credence public key file, created with
.Xr credence 1 .
This file should be owned by the user and have permissions 0600.
It must contain the public part of an Ed25519 key pair.
.It Pa ${HOME}/.credence/known_keys
The user's credence known keys file.
This file must contain the public keys of services the user will access.
Each public key is used to authenticate a remote service during
authentication.
.It Pa <service>/known_keys
A service utilizing libDwmCredence will have a file containing the public keys
of those allowed to access the service.
The location of this file is service dependent.
.El
.Sh SEE ALSO
.Lk .. "Manpage Index"
.Sh AUTHORS
Copyright \(co 2022
.An Daniel W. McRobb Aq Mt dwm@mcplex.net


apps/credence/credence.cc
//===========================================================================
// @(#) $DwmPath$
//===========================================================================
//  Copyright (c) Daniel W. McRobb 2022
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//  1. Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//  2. Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in the
//     documentation and/or other materials provided with the distribution.
//  3. The names of the authors and copyright holders may not be used to
//     endorse or promote products derived from this software without
//     specific prior written permission.
//
//  IN NO EVENT SHALL DANIEL W. MCROBB BE LIABLE TO ANY PARTY FOR
//  DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
//  INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE,
//  EVEN IF DANIEL W. MCROBB HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
//  DAMAGE.
//
//  THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND
//  DANIEL W. MCROBB HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT,
//  UPDATES, ENHANCEMENTS, OR MODIFICATIONS. DANIEL W. MCROBB MAKES NO
//  REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER
//  IMPLIED OR EXPRESS, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE,
//  OR THAT THE USE OF THIS SOFTWARE WILL NOT INFRINGE ANY PATENT,
//  TRADEMARK OR OTHER RIGHTS.
//===========================================================================

//---------------------------------------------------------------------------
//!  \file credence.cc
//!  \author Daniel W. McRobb
//!  \brief credence key generator and checker (Ed25519 keys)
//---------------------------------------------------------------------------

#include "DwmArguments.hh"
#include "DwmCredenceKeyStash.hh"
#include "DwmCredenceVersion.hh"

using namespace std;

typedef   Dwm::Arguments<Dwm::Argument<'i',string>,
                         Dwm::Argument<'d',string>> KeyGenArgType;

typedef   Dwm::Arguments<Dwm::Argument<'d',string>> KeyCheckArgType;

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void Usage(const char *argv0)
{
  cerr << "Usage: " << argv0 << " keygen [-i id] [-d directory]\n"
       << "       " << argv0 << " keycheck [-d directory]\n"
       << "       " << argv0 << " -v\n";
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void InitKeyGenArgs(KeyGenArgType & args)
{
  args.SetValueName<'i'>("identity");
  Dwm::Credence::Ed25519KeyPair  kp;
  args.SetHelp<'i'>("Use the given identity (defaults to " + kp.Id().Value() + ")");
  args.SetValueName<'d'>("directory");
  args.SetHelp<'d'>("directory in which to store keys (defaults to ~/.credence)");
  args.Set<'d'>("~/.credence");
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void InitKeyCheckArgs(KeyCheckArgType & args)
{
  args.SetValueName<'d'>("directory");
  args.SetHelp<'d'>("key directory (defaults to ~/.credence)");
  args.Set<'d'>("~/.credence");
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static bool CheckKeyStash(const string & keyDir)
{
  Dwm::Credence::KeyStash  keyStash(keyDir);
  return keyStash.IsValid();
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
int main(int argc, char *argv[])
{
  KeyCheckArgType  keycheckArgs;
  KeyGenArgType    keygenArgs;
  InitKeyCheckArgs(keycheckArgs);
  InitKeyGenArgs(keygenArgs);

  if (argc < 2) {
    Usage(argv[0]);
    return 1;
  }
  else {
    if (string(argv[1]) == "keygen") {
      int  argind = keygenArgs.Parse(argc-1, &argv[1]);
      if (argind < 0) {
        cerr << keygenArgs.Usage(string(argv[0]) + ' ' + argv[1], "");
        return 1;
      }
      Dwm::Credence::Ed25519KeyPair  keys(keygenArgs.Get<'i'>());
      Dwm::Credence::KeyStash        keyStash(keygenArgs.Get<'d'>());
      if (keyStash.Save(keys)) {
        return 0;
      }
    }
    else if (string(argv[1]) == "keycheck") {
      int  argind = keycheckArgs.Parse(argc-1, &argv[1]);
      if (argind < 0) {
        cerr << keycheckArgs.Usage(string(argv[0]) + ' ' + argv[1], "");
        return 1;
      }
      if (CheckKeyStash(keycheckArgs.Get<'d'>())) {
        cout << "Valid key stash '" << keycheckArgs.Get<'d'>() << "'\n";
        return 0;
      }
      else {
        cerr << "Invalid key stash '" << keycheckArgs.Get<'d'>() << "'\n";
        return 1;
      }
    }
    else if (string(argv[1]) == "-v") {
      cout << Dwm::Credence::Version.Version() << '\n';
      return 0;
    }
    else {
      Usage(argv[0]);
      return 1;
    }
  }
  
  return 1;
}
